jgrapht-core/src/main/java/org/jgrapht/alg/tour/NearestNeighborHeuristicTSP.java

```diff
@@ -0,0 +1,181 @@
+ * </p>
+ * 
+ * <p>
+ * This algorithm requires that the graph is complete.
+ * </p>
+ *
+ * @param <V> the graph vertex type
+ * @param <E> the graph edge type
+ *
+ * @author Peter Harman
+ */
+public class NearestNeighborHeuristicTSP<V, E>
+        extends
+        HamiltonianCycleAlgorithmBase<V, E> {
+
+    private Random rng;
+    private V first;
+
+    /**
+     * Constructor. By default a random vertex is chosen to start.
+     */
+    public NearestNeighborHeuristicTSP() {
+        this(null, new Random());
+    }
+
+    /**
+     * Constructor
+     *
+     * @param first First vertex to visit, or null to choose at random
+     */
+    public NearestNeighborHeuristicTSP(V first) {
+        this(first, new Random());
+    }
+
+    /**
+     * Constructor
+     *
+     * @param seed seed for the random number generator
+     */
+    public NearestNeighborHeuristicTSP(long seed) {
+        this(null, new Random(seed));
+    }
+
+    /**
+     * Constructor
+     *
+     * @param first First vertex to visit, or null to choose at random
+     * @param rng Random number generator
+     */
+    public NearestNeighborHeuristicTSP(V first, Random rng) {
+        this.first = first;
+        this.rng = Objects.requireNonNull(rng, "Random number generator cannot be null");
+    }
+
+    /**
+     * Computes a tour using the nearest neighbour heuristic.
+     *
+     * @param graph the input graph
+     * @return a tour
+     * @throws IllegalArgumentException if the graph is not undirected
+     * @throws IllegalArgumentException if the graph is not complete
+     * @throws IllegalArgumentException if the graph contains no vertices
+     */
+    @Override
+    public GraphPath<V, E> getTour(Graph<V, E> graph) {
+        // Handle a graph with single vertex
+        if (graph.vertexSet().size() == 1) {
+            return getSingletonTour(graph);
+        }
+        Set<V> unvisited = new HashSet<>();
+        // Get the initial vertex
+        V current = first(unvisited, graph);
+        List<V> visited = new ArrayList<>(unvisited.size() + 1);
+        visited.add(current);
+        // Iterate until tour is complete
+        while (!unvisited.isEmpty()) {
+            // Find the nearest vertex and add to the tour
+            current = nearest(current, unvisited, graph);
+            visited.add(current);
+        }
+        return listToTour(visited, graph);
+    }
+
+    /**
+     * Get or determine the first vertex
+     * 
+     * @param unvisited Set to populate with unvisited vertices
+     * @param graph The graph
+     * @return A suitable vertex to start
+     * @throws IllegalArgumentException if the graph is not undirected
+     * @throws IllegalArgumentException if the graph is not complete
+     * @throws IllegalArgumentException if the graph contains no vertices
+     */
+    private V first(Set<V> unvisited, Graph<V, E> graph) {
+        // Check that graph is appropriate
+        checkGraph(graph);
```

Comment:
We should check the graph before we proceed with any actions because this is the behaviour of the existing implementations

Reply:
Agree, done